admiral/pkg/clusters/virtualservice_routing.go

```diff
@@ -0,0 +1,628 @@
+			vsHosts = append(vsHosts, globalFQDN)
+		}
+
+		if len(vsHosts) == 0 {
+			return fmt.Errorf(
+				"skipped creating virtualservice on cluster %s, no valid hosts found", sourceCluster)
+		}
+		if len(tlsRoutes) == 0 {
+			return fmt.Errorf(
+				"skipped creating virtualservice on cluster %s, no valid tls routes found", sourceCluster)
+		}
+		sort.Strings(vsHosts)
+		virtualService.Spec.Hosts = vsHosts
+		sort.Slice(tlsRoutes, func(i, j int) bool {
+			return tlsRoutes[i].Match[0].SniHosts[0] < tlsRoutes[j].Match[0].SniHosts[0]
+		})
+		virtualService.Spec.Tls = tlsRoutes
+
+		// If we were unable to find the default host in the above loop,
+		// then we pick the first one
+		if vsName == "" {
+			vsName = vsHosts[0] + "-routing-vs"
+		}
+		virtualService.Name = vsName
+
+		existingVS, err := getExistingVS(ctxLogger, ctx, rc, virtualService.Name)
+		if err != nil {
+			ctxLogger.Warn(err.Error())
+		}
+
+		ctxLogger.Infof(common.CtxLogFormat, "addUpdateVirtualServicesForSourceIngress",
+			virtualService.Name, virtualService.Namespace, sourceCluster, "Add/Update ingress virtualservice")
+		err = addUpdateVirtualService(
+			ctxLogger, ctx, virtualService, existingVS, common.GetSyncNamespace(), rc, remoteRegistry)
+
+	}
+
+	return nil
+}
+
+// getAllVSRouteDestinationsByCluster generates the route destinations for each source cluster
+// This is during the discovery phase where the route destinations are created for each source cluster
+// For a given identity and env, we are going to build a map of all possible services across deployments
+// and rollouts. This map will be used to create the route destinations for the VirtualService
+func getAllVSRouteDestinationsByCluster(
+	serviceInstance map[string]*k8sV1.Service,
+	meshDeployAndRolloutPorts map[string]map[string]uint32,
+	weightedServices map[string]*WeightedService,
+	rollout *argo.Rollout,
+	deployment *k8sAppsV1.Deployment) (map[string][]*networkingV1Alpha3.RouteDestination, error) {
+
+	if serviceInstance == nil {
+		return nil, fmt.Errorf("serviceInstance is nil")
+	}
+
+	destinations := make(map[string][]*networkingV1Alpha3.RouteDestination)
+	if serviceInstance[common.Deployment] != nil {
+		meshPort, err := getMeshHTTPPortForDeployment(meshDeployAndRolloutPorts)
+		if err != nil {
+			return nil, err
+		}
+		err = populateVSRouteDestinationForDeployment(
+			serviceInstance, meshPort, deployment, destinations)
+		if err != nil {
+			return nil, err
+		}
+	}
+	if serviceInstance[common.Rollout] != nil {
+		meshPort, err := getMeshHTTPPortForRollout(meshDeployAndRolloutPorts)
+		if err != nil {
+			return nil, err
+		}
+		err = populateVSRouteDestinationForRollout(
+			serviceInstance, weightedServices, rollout, meshPort, destinations)
+		if err != nil {
+			return nil, err
+		}
+	}
+	return destinations, nil
+}
+
+// getMeshHTTPPortForRollout gets the mesh http port for the rollout
+func getMeshHTTPPortForRollout(ports map[string]map[string]uint32) (uint32, error) {
```

Comment:
As most of the code is common between this and `getMeshHTTPPortForDeployment` should we just have one function and return the ports for both rollout and deployment?

Also, what are the cases that the ports would be different for rollout and deployment for the same application?